middle_end/flambda2/from_lambda/lambda_to_lambda_transforms.ml

```diff
@@ -110,7 +110,417 @@ let rec_catch_for_for_loop env loc ident start stop
+                 loc ) ),
+         array_layout ))
+
+let makearray_dynamic_scannable_unboxed_product env
+    (lambda_array_kind : L.array_kind) (mode : L.locality_mode) ~length
+    ~(init : L.lambda) loc =
+  let must_be_scanned =
+    match lambda_array_kind with
+    | Pgcignorableproductarray _ -> false
+    | Pgcscannableproductarray kinds ->
+      let rec must_be_scanned (kind : L.scannable_product_element_kind) =
+        match kind with
+        | Pint_scannable -> false
+        | Paddr_scannable -> true
+        | Pproduct_scannable kinds -> List.exists must_be_scanned kinds
+      in
+      List.exists must_be_scanned kinds
+    | Pgenarray | Paddrarray | Pintarray | Pfloatarray | Punboxedfloatarray _
+    | Punboxedintarray _ | Punboxedvectorarray _ ->
+      Misc.fatal_errorf
+        "%s: should have been sent to [makearray_dynamic_singleton]"
+        (Printlambda.array_kind lambda_array_kind)
+  in
+  if must_be_scanned
+  then
+    ( env,
+      makearray_dynamic_scannable_unboxed_product0 lambda_array_kind mode
+        ~length ~init loc )
+  else
+    makearray_dynamic_non_scannable_unboxed_product env lambda_array_kind mode
+      ~length ~init:(Some init) loc
+
+let makearray_dynamic env (lambda_array_kind : L.array_kind)
+    (mode : L.locality_mode) args loc : Env.t * primitive_transform_result =
+  (* %makearray_dynamic is analogous to (from stdlib/array.ml):
+   *   external create: int -> 'a -> 'a array = "caml_make_vect"
+   * except that it works on any layout, including unboxed products, at both
+   * heap and local modes.
+   * Additionally, if the initializer is omitted, an uninitialized array will
+   * be returned.  Initializers must however be provided when the array kind is
+   * Pgenarray, Paddrarray, Pintarray, Pfloatarray or Pgcscannableproductarray;
+   * or when a Pgcignorablearray involves an [int].  (See comment below.)
+   *)
+  let dbg = Debuginfo.from_location loc in
+  let length, init =
+    match args with
+    | [length] -> length, None
+    | [length; init] -> length, Some init
+    | [] | _ :: _ :: _ ->
+      Misc.fatal_error
+        "%makearray_dynamic takes the (non-unarized) length and optionally an \
+         initializer (the latter perhaps of unboxed product layout)"
+  in
+  let[@inline] must_have_initializer () =
+    match init with
+    | Some init -> init
+    | None -> (
+      match lambda_array_kind with
+      | Pintarray ->
+        Misc.fatal_errorf
+          "Cannot compile Pmakearray_dynamic at layout Pintarray without an \
+           initializer; otherwise it might be possible for values of type \
+           [int] having incorrect representations to be revealed, thus \
+           breaking soundness:@ %a"
+          Debuginfo.print_compact dbg
+      | Pgenarray | Paddrarray | Pfloatarray | Punboxedfloatarray _
+      | Punboxedintarray _ | Punboxedvectorarray _ | Pgcscannableproductarray _
+      | Pgcignorableproductarray _ ->
```

Comment:
`Pgcignorableproductarray` involving an int should also have a similar error to above